blog/content/post/2024-03-15-uefi-sb-implementation-mpl.md

```diff
@@ -0,0 +1,385 @@
+system's boot partition layout:
+
+```conf
+IMAGE_BOOT_FILES = " \
+  bootfiles/EFI/BOOT/grub*;EFI/BOOT/ \
+  bootfiles/EFI/BOOT/x86_64-efi/*;EFI/BOOT/x86_64-efi/ \
+  bzImage-initramfs-${MACHINE}.bin;bzImage-initramfs-a \
+  bzImage-initramfs-${MACHINE}.bin;bzImage-initramfs-b \
+  bzImage-initramfs-${MACHINE}.bin.sig;bzImage-initramfs-a.sig \
+  bzImage-initramfs-${MACHINE}.bin.sig;bzImage-initramfs-b.sig \
+  bootx64.efi;EFI/BOOT/ \
+"
+```
+
+At this point, we encountered several problems with our build.
+
+With `GRUB_SIGN_VERIFY` variable enabled, every grub component needed to be
+signed so that grub could use it. The recipes from `meta-efi-secure-boot`
+take
+care of generating the signatures. However, the `grubenv` signature was missing
+from our output files. As it turns out, the layer does not automatically sign
+that file. We had to append that feature to the `grub-efi` recipe:
+
+```bitbake
+# grub-efi_%.bbappend
+
+fakeroot python do_sign:append:class-target() {
+    uks_bl_sign(dir + 'grubenv', d)
+}
+
+fakeroot do_chownboot:append() {
+    chown root:root -R "${D}${EFI_BOOT_PATH}/grubenv${SB_FILE_EXT}"
+}
+
+do_deploy:append:class-target () {
+    # deploy missing grubenv.sig file
+    install -m 0600 "${D}${EFI_BOOT_PATH}/grubenv${SB_FILE_EXT}" "${DEPLOYDIR}"
+}
+```
+
+Another issue was that poky has its own `grub-efi` recipe, where it installs its
+own `grub-efi-grubx64.efi` file. It conflicted with the boot files from
+`meta-efi-secure-boot`, so we had to remove it:
+
+```bitbake
+# grub-efi_%.bbappend
+
+do_deploy:append:class-target () {
+    # remove default grub-efi-grubx64.efi file deployed by .bb from poky; when
+    # it is left in deploydir, bootimg-efi plugin picks it up after cloning
+    # files from IMAGE_BOOT_FILES list
+    # see: https://git.yoctoproject.org/poky/tree/scripts/lib/wic/plugins/source/bootimg-efi.py?id=00c04394cbc5ecaced7cc1bc8bc8787e621f987d#n360
+    rm -rf ${DEPLOYDIR}/${GRUB_IMAGE_PREFIX}${GRUB_IMAGE}
+}
+```
+
+Now your files should be signed automatically during the build. They will be
+deployed along with their signatures. By default, they will be signed using the
+sample keys from
+[meta-signing-key](https://github.com/Wind-River/meta-secure-core/tree/master/meta-signing-key/files/uefi_sb_keys)
+, which is of course extremely unsafe and should only be used for testing.
+`meta-signing-key` provides a
+[script](https://github.com/Wind-River/meta-secure-core/blob/master/meta-signing-key/scripts/create-user-key-store.sh)
+, which generates custom user keys. Such keys should be enrolled in UEFI so
+that Secure Boot can be safely utilized.
+
+## Summary
+
+We managed to show that utilizing UEFI Secure Boot on MPL PIP4x platforms is
+feasible, although some features, such as verifying that the certificate used to
+sign bootable files has not expired, are not supported. The `meta-secure-core`
+layer helps developers implement automatic file signing and verification with
+relative ease.
```

Comment:
@PLangowski here we should market all products related to this blog post:
- Compatibility report for UEFI Secure Boot integration
- UEFI Secure Boot Yocto build integration
- Automated device provisioning for UEFI Secure Boot
- we can also mention what can be done after correctly implementing UEFI SB

@artur-rs @macpijan let's make sure this campaign will be executed correctly. That means there will be products in the shop to which we can link. @rafkoch @Stanislaw-bnk cc

Reply:
I wrote about the first 3 points. @macpijan Do you have some ideas about how I could tackle the last point?

Reply:
CI/CD for automatic signing? HSM or other method for secure signing may storage.

Reply:
Signing is done during Yocto build, so you can't really do it in CI, unless you mean automatically building the image. If I understand correctly, the case is similar with HSM. Maybe storing keys would work here.

Reply:
> Signing is done during Yocto build, so you can't really do it in CI, unless you mean automatically building the image.

I need help understanding this argument. CI/CD is precisely used to build artifacts from the given code. In this case, building Zarhus' image, since the singing process is part of creating an image, then the results of CI/CD would be a signed image. That is very good and worth mentioning, but more is needed to answer what we can do after delivering the products mentioned in my first comment.

So the critical question is where Zarhus will take the keys required for signing and what options we have to store those keys. This is precisely what I'm asking here: potential improvements and next steps. Either we know how to have a fully-featured CI/CD pipeline for producing Zarhus images ready for deployment on the UEFI Secure Boot-enabled platform, or things require further development, which could be described in the post. The critical point is that we answered the following considerations: Yes, we got through the report, we did integration, and we have automatic device provisioning. What do you think next should be? Next could be CI/CD integration, automatic signing, and preparation of required HSM or another secure critical storage method, so in the future, any updates and changes to the image will be handled automatically. The next question could be: How can we update this system so that it does preserve the security properties we provided?

Please let me know if you understand what I'm asking for.

Also, wherever we add value (there is custom code or improvement of existing), I would appreciate it if you would use Zarhus instead of Yocto to ensure we market our solution.

Reply:
Yes, I think I understand. I added a section, where I desctibe the possible steps we can take next, as well as their feasibility and potential risks. Let me know if this looks okay to you. e42216e38453cebc2a79bae9c921400df9c09445

Reply:
@PLangowski, I like your description. What I would add is some issue [here](https://github.com/zarhus/zarhus-issues/issues) briefly describing the problem and need for PKCS#11 used in the Yocto build process for secure usage of HSM. This could be a great conference talk and/or product topic, so it would be great if we would not lose the idea. After submitting the issue and linking it here, I'm ok with resolving this conversation.

Reply:
Thank you. Also, I changed a few instances of the word `Yocto` into `Zarhus` where it made sense: 50978cce2d3f6125b04b7988412af360e0aea670

Reply:
I created the issue. https://github.com/zarhus/zarhus-issues/issues/3

Reply:
Thank you